### PR TITLE
Maple Integration (Minting)

### PR DESCRIPTION
**Pull request by crypto-vincent**
This PR adds support for maple depository new instructions : register, mint and edit

PR Combo:
- uxd-client: https://github.com/UXDProtocol/uxd-client/pull/29
- uxd-program: https://github.com/UXDProtocol/uxd-program/pull/205

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -262,6 +266,103 @@ export class UXDClient {
...
+          mapleLenderShares: depository.mapleLenderShares,
+          systemProgram: SystemProgram.programId,
+          associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
+          tokenProgram: TOKEN_PROGRAM_ID,
+          syrupProgram: depository.syrupProgramId,
+          rent: SYSVAR_RENT_PUBKEY,
+        },
+        options,
+      }
+    );
```
Comment (Orelsanpls) [inline src/client.ts:363]: You can use property shorthand syntax when the name of the property is the same as the variable. Also can use destructuration to simplify? (this one is opiniated)

```suggestion
    const {
          depositoryCollateral,
          maplePool,
          maplePoolLocker,
          mapleGlobals,
          mapleLender,
          mapleSharesMint,
          mapleLockedShares,
          mapleLenderShares,
    } = depository;
   
    return this.instruction.mintWithMaplePoolDepository(
      nativeCollateralAmount,
      {
        accounts: {
          user: authority,
          payer: payer ?? authority,
          controller: controller.pda,
          depository: depository.pda,
          depositoryCollateral,
          redeemableMint,
          userRedeemable: userRedeemableATA,
          collateralMint,
          userCollateral: userCollateralATA,
          maplePool,
          maplePoolLocker,
          mapleGlobals,
          mapleLender,
          mapleSharesMint,
          mapleLockedShares,
          mapleLenderShares,
          systemProgram: SystemProgram.programId,
          associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
          tokenProgram: TOKEN_PROGRAM_ID,
          syrupProgram: depository.syrupProgramId,
          rent: SYSVAR_RENT_PUBKEY,
        },
        options,
      }
    );
```

File: src/maple_pool/depository.ts [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    );
+
+    const mapleLockedShares = await this.findLockedSharesAddress(
+      mapleLender,
+      syrupProgramId
+    );
+    const mapleLenderShares = await this.findLenderSharesAddress(
+      depository,
+      mapleSharesMint
+    );
```
Comment (Orelsanpls) [inline src/maple_pool/depository.ts:115]: May be interesting to use `Promise.all` to boost the speed a bit

```suggestion
    const [
        collateralInfo,
        depositoryCollateral,
        mapleLender,
        mapleLenderShares,
    ] = await Promise.all([
        collateralToken.getMintInfo(),
        
        this.findDepositoryCollateralAddress(
            depository,
            uxdProgramId
        ),

        this.findLenderAddress(
            maplePool,
            depository,
            syrupProgramId
        ),

        this.findLenderSharesAddress(
            depository,
            mapleSharesMint
        ),
    ]);

    if (!collateralInfo) {
      throw new Error('Cannot find the collateral mint');
    }
    const {
        decimals: collateralDecimals,
    } = collateralInfo;
    
    const mapleLockedShares = await this.findLockedSharesAddress(
      mapleLender,
      syrupProgramId
    );
```
```

File: src/maple_pool/depository.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+  private static async findDepositoryCollateralAddress(
+    depository: PublicKey,
+    uxdProgramId: PublicKey
+  ) {
+    return (
+      await PublicKey.findProgramAddress(
+        [
+          Buffer.from(MAPLE_POOL_DEPOSITORY_COLLATERAL_NAMESPACE),
+          depository.toBytes(),
```
Comment (Orelsanpls) [inline src/maple_pool/depository.ts:142]: We usually use `toBuffer()`:

```suggestion
          depository.toBuffer(),
```

No diff between `toBuffer` and `toBytes`.

```
    /**
     * Return the byte array representation of the public key
     */
    toBytes(): Uint8Array;
    /**
     * Return the Buffer representation of the public key
     */
    toBuffer(): Buffer;
```

```
    static findProgramAddressSync(
      seeds: Array<Buffer | Uint8Array>,
      programId: PublicKey,
    ): [PublicKey, number];
```

File: src/maple_pool/depository.ts [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    maplePool: PublicKey,
+    depository: PublicKey,
+    syrupProgramId: PublicKey
+  ): Promise<PublicKey> {
+    return (
+      await PublicKey.findProgramAddress(
+        [
+          Buffer.from(MAPLE_INTERNAL_LENDER_NAMESPACE),
+          maplePool.toBytes(),
+          depository.toBytes(),
```
Comment (Orelsanpls) [inline src/maple_pool/depository.ts:159]: Same:

```suggestion
          maplePool.toBuffer(),
          depository.toBuffer(),
```

File: src/maple_pool/depository.ts [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+  private static async findLockedSharesAddress(
+    mapleLender: PublicKey,
+    syrupProgramId: PublicKey
+  ): Promise<PublicKey> {
+    return (
+      await PublicKey.findProgramAddress(
+        [
+          Buffer.from(MAPLE_INTERNAL_LOCKED_SHARES_NAMESPACE),
+          mapleLender.toBytes(),
```
Comment (Orelsanpls) [inline src/maple_pool/depository.ts:181]: same

```suggestion
          mapleLender.toBuffer(),
```

File: src/maple_pool/depository.ts [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+      ['collateralDecimals']: this.collateralDecimals,
+      ['collateralSymbol']: this.collateralSymbol,
+      ['maplePool']: this.maplePool.toBase58(),
+      ['maplePoolLocker']: this.maplePoolLocker.toBase58(),
+      ['mapleGlobals']: this.mapleGlobals.toBase58(),
+      ['mapleLender']: this.mapleLender.toBase58(),
+      ['mapleSharesMint']: this.mapleSharesMint.toBase58(),
+      ['mapleLockedShares']: this.mapleLockedShares.toBase58(),
+      ['mapleLenderShares']: this.mapleLenderShares.toBase58(),
+    });
```
Comment (Orelsanpls) [inline src/maple_pool/depository.ts:203]: hm, I just see it now kek _(have copy pasted it as well in mercurial)_

actually there are no reason for the keys to be wrapped lol

```suggestion
    console.table({
      pda: this.pda.toBase58(),
      depositoryCollateral: this.depositoryCollateral.toBase58(),
      collateralMint: this.collateralMint.toBase58(),
      collateralDecimals: this.collateralDecimals,
      collateralSymbol: this.collateralSymbol,
      maplePool: this.maplePool.toBase58(),
      maplePoolLocker: this.maplePoolLocker.toBase58(),
      mapleGlobals: this.mapleGlobals.toBase58(),
      mapleLender: this.mapleLender.toBase58(),
      mapleSharesMint: this.mapleSharesMint.toBase58(),
      mapleLockedShares: this.mapleLockedShares.toBase58(),
      mapleLenderShares: this.mapleLenderShares.toBase58(),
    });
```

**Review by Orelsanpls — COMMENTED** [review]
let's go!

**Comment by crypto-vincent** [comment]
Pausing maple integration efforts for now

**Pull request closed without merging**